// @taqueria/protocol/Contract.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// Contract.ts
var Contract_exports = {};
__export(Contract_exports, {
  create: () => create,
  internalSchema: () => internalSchema,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema2,
  schemas: () => schemas2
});
module.exports = __toCommonJS(Contract_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var SHA256 = __toESM(require("@taqueria/protocol/SHA256"));
var import_zod = require("zod");
var rawSchema2 = import_zod.z.object({
  sourceFile: import_zod.z.string().min(1),
  hash: SHA256.rawSchema
});
var internalSchema = import_zod.z.object({
  sourceFile: import_zod.z.string().min(1),
  hash: SHA256.schemas.schema
});
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema,
  parseErrMsg: (value) => `${value} is not a valid contract`,
  unknownErrMsg: "Something went wrong trying to parse a contract"
});
var { create, of, make } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  internalSchema,
  make,
  of,
  rawSchema,
  schemas
});
