// @taqueria/protocol/LoadedConfig.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// LoadedConfig.ts
var LoadedConfig_exports = {};
__export(LoadedConfig_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  internalSchema: () => internalSchema2,
  make: () => make2,
  of: () => of,
  rawSchema: () => rawSchema4,
  schemas: () => schemas3,
  toConfig: () => toConfig
});
module.exports = __toCommonJS(LoadedConfig_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var Config = __toESM(require("@taqueria/protocol/Config"));
var SanitizedAbsPath = __toESM(require("@taqueria/protocol/SanitizedAbsPath"));
var SHA256 = __toESM(require("@taqueria/protocol/SHA256"));
var rawSchema4 = Config.rawSchema.omit({ plugins: true }).extend({
  plugins: Config.pluginsRawSchema,
  projectDir: SanitizedAbsPath.rawSchema.describe("loadedConfig.projectDir"),
  configFile: SanitizedAbsPath.rawSchema.describe("loadedConfig.configFile"),
  hash: SHA256.rawSchema.describe("loadedConfig.hash")
}).describe("LoadedConfig");
var internalSchema2 = Config.internalSchema.omit({ plugins: true }).extend({
  plugins: Config.pluginsInternalSchema,
  projectDir: SanitizedAbsPath.schemas.schema.describe("loadedConfig.projectDir"),
  configFile: SanitizedAbsPath.schemas.schema.describe("loadedConfig.configFile"),
  hash: SHA256.schemas.schema.describe("loadedConfig.hash")
}).describe("LoadedConfig");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema4,
  internalSchema: internalSchema2,
  parseErrMsg: (value) => `The following configuration is invalid: ${value}`,
  unknownErrMsg: "Something went wrong trying to parse the configuration to load"
});
var { create, of, make: make2 } = factory;
var schemas3 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
var toConfig = (config) => Config.make(config);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  internalSchema,
  make,
  of,
  rawSchema,
  schemas,
  toConfig
});
