// @taqueria/protocol/NetworkConfig.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// NetworkConfig.ts
var NetworkConfig_exports = {};
__export(NetworkConfig_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema5,
  schemas: () => schemas5
});
module.exports = __toCommonJS(NetworkConfig_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var EconomicalProtocolHash = __toESM(require("@taqueria/protocol/EconomicalProtocolHash"));
var Faucet = __toESM(require("@taqueria/protocol/Faucet"));
var HumanReadableIdentifier = __toESM(require("@taqueria/protocol/HumanReadableIdentifier"));
var Url = __toESM(require("@taqueria/protocol/Url"));
var import_zod = require("zod");
var rawSchema5 = import_zod.z.object({
  label: HumanReadableIdentifier.rawSchema,
  rpcUrl: Url.rawSchema,
  protocol: EconomicalProtocolHash.rawSchema,
  accounts: import_zod.z.record(
    import_zod.z.any({ description: "Accounts" }),
    { description: "Accounts" }
  ).optional(),
  faucet: Faucet.rawSchema.describe("Network Faucet").optional()
}).describe("Network Config");
var internalSchema = import_zod.z.object({
  label: HumanReadableIdentifier.schemas.schema.describe("Network Label"),
  rpcUrl: Url.schemas.schema.describe("Network RPC Url"),
  protocol: EconomicalProtocolHash.schemas.schema.describe("Network Protocol Hash"),
  accounts: import_zod.z.record(
    import_zod.z.any({ description: "Accounts" }),
    { description: "Accounts" }
  ).optional(),
  faucet: Faucet.schemas.schema.describe("Network Faucet").optional()
}).describe("Network Config");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema5,
  internalSchema,
  parseErrMsg: (value) => `${value} is not a valid network configuration`,
  unknownErrMsg: "Something went wrong trying to parse the network configuration"
});
var { create, of, make } = factory;
var schemas5 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  make,
  of,
  rawSchema,
  schemas
});
