// @taqueria/protocol/ParsedTemplate.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// ParsedTemplate.ts
var ParsedTemplate_exports = {};
__export(ParsedTemplate_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema2,
  schemas: () => schemas2
});
module.exports = __toCommonJS(ParsedTemplate_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var PluginResponseEncoding = __toESM(require("@taqueria/protocol/PluginResponseEncoding"));
var Template = __toESM(require("@taqueria/protocol/Template"));
var import_zod = require("zod");
var internalSchema2 = Template.internalSchema.extend({
  handler: import_zod.z.string(),
  encoding: PluginResponseEncoding.schemas.schema
}).describe("ParsedTemplate");
var rawSchema2 = Template.rawSchema.extend({
  handler: import_zod.z.string(),
  encoding: PluginResponseEncoding.schemas.schema
}).describe("ParsedTemplate");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema: internalSchema2,
  parseErrMsg: (value) => `Could not parse the following template: ${value}`,
  unknownErrMsg: "Something went wrong trying to parse an template"
});
var { create, make, of } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  make,
  of,
  rawSchema,
  schemas
});
