// @taqueria/protocol/PluginDependenciesResponse.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// PluginDependenciesResponse.ts
var PluginDependenciesResponse_exports = {};
__export(PluginDependenciesResponse_exports, {
  create: () => create,
  factory: () => factory,
  make: () => make,
  of: () => of,
  schemas: () => schemas
});
module.exports = __toCommonJS(PluginDependenciesResponse_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var import_zod = require("zod");
var rawRuntimeDependency = import_zod.z.object({
  name: import_zod.z.string(),
  path: import_zod.z.string(),
  version: import_zod.z.string(),
  kind: import_zod.z.union([import_zod.z.literal("required"), import_zod.z.literal("optional")])
});
var { schemas, factory } = (0, import_Base.default)({
  rawSchema: rawRuntimeDependency,
  internalSchema: rawRuntimeDependency,
  parseErrMsg: (value) => `The following runtime dependency is invalid: ${value}`,
  unknownErrMsg: "Something went wrong trying to parse the template"
});
var runtimeDependencyReport = rawRuntimeDependency.extend({ met: import_zod.z.boolean() });
var dependenciesResponseSchema = import_zod.z.object({
  report: import_zod.z.array(runtimeDependencyReport)
});
var { of, make, create } = factory;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  make,
  of,
  schemas
});
