// @taqueria/protocol/PluginProxyResponse.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// PluginProxyResponse.ts
var PluginProxyResponse_exports = {};
__export(PluginProxyResponse_exports, {
  create: () => create,
  factory: () => factory,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema,
  schemas: () => schemas2
});
module.exports = __toCommonJS(PluginProxyResponse_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var PluginJsonResponse = __toESM(require("@taqueria/protocol/PluginJsonResponse"));
var import_zod = require("zod");
var internalSchema = import_zod.z.union([
  PluginJsonResponse.schemas.internalSchema,
  import_zod.z.void()
]);
var rawSchema = internalSchema;
var { schemas: schemas2, factory } = (0, import_Base.default)({
  rawSchema,
  internalSchema,
  parseErrMsg: "The response from the proxy request is invalid",
  unknownErrMsg: "Something went wrong parsing the proxy response from the plugin"
});
var { of, make, create } = factory;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  make,
  of,
  rawSchema,
  schemas
});
