// @taqueria/protocol/PluginResponseEncoding.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// PluginResponseEncoding.ts
var PluginResponseEncoding_exports = {};
__export(PluginResponseEncoding_exports, {
  create: () => create,
  factory: () => factory,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema,
  schemas: () => schemas
});
module.exports = __toCommonJS(PluginResponseEncoding_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var import_zod = require("zod");
var rawSchema = import_zod.z.preprocess(
  (val) => val ?? "none",
  import_zod.z.union([
    import_zod.z.literal("json"),
    import_zod.z.literal("application/json"),
    import_zod.z.literal("none")
  ]).default("none").optional()
);
var { schemas, factory } = (0, import_Base.default)({
  rawSchema,
  internalSchema: rawSchema,
  parseErrMsg: (value) => `The following encoding is invalid: ${value}`,
  unknownErrMsg: "Something went wrong trying to parse the encoding"
});
var { create, of, make } = factory;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  make,
  of,
  rawSchema,
  schemas
});
