// @taqueria/protocol/PositionalArg.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// PositionalArg.ts
var PositionalArg_exports = {};
__export(PositionalArg_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema2,
  schemas: () => schemas2
});
module.exports = __toCommonJS(PositionalArg_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var HumanReadableIdentifier = __toESM(require("@taqueria/protocol/HumanReadableIdentifier"));
var import_zod = require("zod");
var rawSchema2 = import_zod.z.object({
  placeholder: HumanReadableIdentifier.rawSchema.describe("Positional Arg Placeholder"),
  description: import_zod.z.string({ description: "Positional Arg Description" }).min(1),
  defaultValue: import_zod.z.union(
    [import_zod.z.string(), import_zod.z.number(), import_zod.z.boolean()],
    { description: "Positional Arg Default Value" }
  ).optional(),
  type: import_zod.z.union(
    [import_zod.z.literal("string"), import_zod.z.literal("number"), import_zod.z.literal("boolean")],
    { description: "Positional Arg Datatype" }
  ).optional(),
  required: import_zod.z.boolean().optional()
}).describe("Positional Arg");
var internalSchema = rawSchema2.extend({
  placeholder: HumanReadableIdentifier.schemas.schema.describe("Positional Arg Placeholder")
}).describe("Positional Arg");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema,
  parseErrMsg: (value) => `The following positional argument is invalid: ${value}`,
  unknownErrMsg: "Something went wrong parsing the positional argument"
});
var { create, of, make } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  make,
  of,
  rawSchema,
  schemas
});
