// @taqueria/protocol/Provisions.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// Provisions.ts
var Provisions_exports = {};
__export(Provisions_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  schemas: () => schemas2
});
module.exports = __toCommonJS(Provisions_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var Provisioner = __toESM(require("@taqueria/protocol/Provisioner"));
var import_rambda = require("rambda");
var import_rambdax = require("rambdax");
var import_zod = require("zod");
var getInvalidIds = (0, import_rambdax.memoize)(
  (provisions) => {
    const ids = provisions.map((p) => p.id);
    return provisions.reduce(
      (retval, provision) => {
        const depends_on = provision.depends_on ?? [];
        const results = (0, import_rambdax.partition)(
          (id) => ids.includes(id),
          depends_on
        );
        const invalid = results.pop();
        return (0, import_rambda.uniq)([...retval, ...invalid]);
      },
      []
    );
  }
);
var rawSchema2 = import_zod.z.array(
  Provisioner.rawSchema
).refine(
  (provisions) => getInvalidIds(provisions).length === 0,
  (provisions) => ({
    message: `One or more of your provisioners depends on an invalid provisioner. The following provisioner ids were referenced that do not exist: ${getInvalidIds(provisions).join(", ")}`
  })
).describe("Provisions");
var internalSchema = import_zod.z.array(Provisioner.schemas.schema).refine(
  (provisions) => getInvalidIds(provisions).length === 0,
  (provisions) => ({
    message: `One or more of your provisioners depends on an invalid provisioner. The following provisioner ids were referenced that do not exist: ${getInvalidIds(provisions).join(", ")}`
  })
).describe("Provisions");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema,
  parseErrMsg: (value) => `The following provision is invalid: ${value}`,
  unknownErrMsg: "Something went wrong parsing the list of provisioners"
});
var { create, of, make } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  make,
  of,
  schemas
});
