// @taqueria/protocol/Provisioner.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// Provisioner.ts
var Provisioner_exports = {};
__export(Provisioner_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  internalSchema: () => internalSchema,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema2,
  schemas: () => schemas2
});
module.exports = __toCommonJS(Provisioner_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var ProvisionerID = __toESM(require("@taqueria/protocol/ProvisionerID"));
var import_zod = require("zod");
var rawSchema2 = import_zod.z.object(
  {
    id: ProvisionerID.rawSchema,
    plugin: import_zod.z.string().min(1).describe("Provisioner Plugin").optional(),
    operation: import_zod.z.union([
      import_zod.z.string().min(1),
      import_zod.z.literal("custom")
    ]).describe("Provisioner Operation"),
    command: import_zod.z.string().describe("Provisioner Custom Command").optional(),
    label: import_zod.z.string().describe("Provisioner Label").optional(),
    depends_on: import_zod.z.array(
      ProvisionerID.rawSchema
    ).optional()
  },
  {
    description: "Provisioner"
  }
).passthrough();
var internalSchema = rawSchema2.extend({
  id: ProvisionerID.schemas.schema,
  depends_on: import_zod.z.array(ProvisionerID.schemas.schema).optional()
});
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema,
  parseErrMsg: "There was a problem trying to parse the provisioner",
  unknownErrMsg: "Something went wrong trying to parse the provisioner"
});
var { create, of, make } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  create,
  factory,
  generatedSchemas,
  internalSchema,
  make,
  of,
  rawSchema,
  schemas
});
